File: architecture/src/main/java/com/android/architecture/utils/AsyncTask.java
package com.android.architecture.utils;

import com.android.architecture.helper.AppExecutors;

/**
 * File describe:
 * Author: SuQi
 * Create date: 2022/7/29
 * Modify date: 2022/7/29
 * Version: 1
 */
public class AsyncTask {

    public static <T> void doAction(ActionStart<T> start, ActionEnd<T> end) {
        AppExecutors.getInstance().io().execute(() -> {
            T data = start.getData();
            AppExecutors.getInstance().main().execute(() -> {
                end.onResult(data);
            });
        });
    }

    public interface ActionStart<T> {
        T getData();
    }

    public interface ActionEnd<T> {
        void onResult(T t);
    }

}

File: architecture/src/main/java/com/android/architecture/helper/AsyncTask.kt
package com.android.architecture.helper

/**
 * File describe:异步任务处理
 * Author: SuQi
 * Create date: 2022/10/21
 * Modify date: 2022/10/21
 * Version: 1
 */
object AsyncTask {

    fun <T> execute(start: () -> T, end: ((result: T) -> Unit)? = null) {
        AppExecutors.getInstance().io().execute {
            val result = start.invoke()
            AppExecutors.getInstance().main().execute {
                end?.invoke(result)
            }
        }
    }

}